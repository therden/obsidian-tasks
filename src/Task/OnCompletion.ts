import { TFile } from 'obsidian';

import { StatusType } from '../Statuses/StatusConfiguration';
import { appendToEndOfFile } from '../lib/FileWriter';
import type { Task } from './Task';

function removeCalloutPrefixes(lineOfText: string) {
    while (lineOfText.substring(0, 2) === '> ') {
        lineOfText = lineOfText.substring(2);
    }
    return lineOfText;
}

function returnWithoutCompletedInstance(tasks: Task[], changedStatusTask: Task) {
    return tasks.filter((task) => task !== changedStatusTask);
}

async function updateFileContent(filePath: string, fileContentUpdater: (data: string) => string): Promise<void> {
    let file = app.vault.getAbstractFileByPath(filePath);
    if (file === null) {
        // Try creating the file.
        // This probably depends on any parent directories already existing:
        // TODO If filePath is not in root, if necessary, create intermediate directories.
        file = await app.vault.create(filePath, '');
    }

    if (file instanceof TFile) {
        await app.vault.process(file, (data) => {
            return fileContentUpdater(data);
        });
    } else {
        // If we were not able to save the done task, we would like to be able to retain everything.
        // TODO There is currently no way to communicate this failure back to callers,
        //      and so if we reach here, the completed task gets unintentionally discarded.
        console.log(`Something went wrong - cannot read or create ${filePath}`);
    }
}

export function updateFileContentEventually(filePath: string, fileContentUpdater: (data: string) => string): void {
    updateFileContent(filePath, fileContentUpdater).then(() => {});
}

export function adjustLineForArchival(changedStatusTaskLine: string) {
    return removeCalloutPrefixes(changedStatusTaskLine.trimStart());
}

export function handleOnCompletion(
    task: Task,
    tasks: Task[],
    toLogFilePath: string = 'Manual Testing/On Completion/Archive.md',
    fileWriter: (filePath: string, fileContentUpdater: (data: string) => string) => void = updateFileContentEventually,
): Task[] {
    const tasksArrayLength = tasks.length;
    if (tasksArrayLength === 0) {
        return tasks;
    }
    const startStatus = task.status;

    const changedStatusTask = tasks[tasksArrayLength - 1];
    const endStatus = changedStatusTask.status;

    const ocTrigger = ' 🏁 ';
    const taskString = changedStatusTask.description;

    if (!taskString.includes(ocTrigger) || endStatus.type !== StatusType.DONE || endStatus.type === startStatus.type) {
        return tasks;
    }

    if (taskString.includes('🏁 Delete')) {
        return returnWithoutCompletedInstance(tasks, changedStatusTask);
    }

    // trim leading spaces and remove `> ` prefixes to prevent misinterpretation
    //    of completed task instances s moved to new contexts
    const changedStatusTaskLine = changedStatusTask.toFileLineString();
    const textToWrite: string = adjustLineForArchival(changedStatusTaskLine);

    if (taskString.includes('🏁 ToLogFile')) {
        //  append completed task to end of list under specified heading of separate, specified note file
        fileWriter(toLogFilePath, (data: string) => {
            return appendToEndOfFile(data, textToWrite);
        });
        return returnWithoutCompletedInstance(tasks, changedStatusTask);
    }

    if (taskString.includes('🏁 ToLogList')) {
        //  move completed task to end of list with specified heading within note in which it originated
        const filePath = changedStatusTask.path;
        fileWriter(filePath, (data: string) => {
            // return appendToListWithinFile(data, '## Archived Tasks - Prepended', textToWrite);
            return writeLineToListEnd(data, '## Archived Tasks - Appended', textToWrite);
        });
        return returnWithoutCompletedInstance(tasks, changedStatusTask);
    }

    if (taskString.includes('🏁 EndOfList')) {
        //  move completed task to end of list in which it originated
        const filePath = changedStatusTask.path;
        fileWriter(filePath, (data: string) => {
            // TODO The function name says that it writes to the end of the list, but it writes to the start.
            // TODO It does not create the heading if it was missing.
            return writeLineToListEnd(data, '## Archived Tasks - Appended', textToWrite);
        });
        return returnWithoutCompletedInstance(tasks, changedStatusTask);
    }

    // const errorMessage = 'Unknown "On Completion" action: ' + ocAction;
    const errorMessage = 'Unknown "On Completion" action';
    console.log(errorMessage);
    return tasks;
    // const hint = '\nClick here to clear';
    // const noticeMessage = errorMessage + hint;
    // new Notice(noticeMessage, 0);
    // console.log('Uh-oh -- we should never actually get here...  :( ');
    // throw new Error('Something went wrong');
}

export function writeLineToListEnd(
    initialFileContent: string,
    targetListHeading: string,
    textToAppend: string | any[],
) {
    if (textToAppend.length === 0) {
        return initialFileContent;
    }
    if (targetListHeading === '') {
        throw Error('Cannot move line to list as empty target list heading was supplied');
    }
    const NEWLINE = '\n';
    const TASK_REGEX = new RegExp('^(> )*( *(- \\[.]))');
    const linesArray = initialFileContent.split('\n');
    let insertionLine = linesArray.indexOf(targetListHeading) + 1;
    for (const thisLine of linesArray.slice(insertionLine)) {
        if (thisLine.search(TASK_REGEX) === -1) break;
        insertionLine += 1;
    }
    linesArray[insertionLine - 1] += NEWLINE + textToAppend;
    return linesArray.join(NEWLINE);
}
