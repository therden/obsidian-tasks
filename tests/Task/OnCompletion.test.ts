/**
 * @jest-environment jsdom
 */
import moment from 'moment';
import { verifyAll } from 'approvals/lib/Providers/Jest/JestApprovals';
import { RecurrenceBuilder } from '../TestingTools/RecurrenceBuilder';
import { Status } from '../../src/Statuses/Status';
import { StatusConfiguration, StatusType } from '../../src/Statuses/StatusConfiguration';
import { TaskBuilder } from '../TestingTools/TaskBuilder';
import { fromLine, toLines, toMarkdown } from '../TestingTools/TestHelpers';
import type { Task } from '../../src/Task/Task';
import { handleOnCompletion } from '../../src/Task/OnCompletion';
import { writeLineToListEnd } from '../../src/Task/OnCompletion';

window.moment = moment;

beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date('2024-02-11'));
});

afterEach(() => {
    jest.useRealTimers();
    // resetSettings();
});

export function applyStatusAndOnCompletionAction(task: Task, newStatus: Status) {
    const tasks = task.handleNewStatus(newStatus);
    function testFileWriter(_filePath: string, _fileContentUpdater: (data: string) => string): void {
        // updateFileContent(filePath, fileContentUpdater).then(() => {});
    }
    return handleOnCompletion(task, tasks, 'archive.md', testFileWriter);
}

describe('OnCompletion', () => {
    it('should just return task if Action is not recognized', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const task = new TaskBuilder()
            .description('A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION')
            .dueDate(dueDate)
            .build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const returnedTasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(returnedTasks.length).toEqual(1);
        expect(toLines(returnedTasks).join('\n')).toMatchInlineSnapshot(
            '"- [x] A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION 📅 2024-02-10 ✅ 2024-02-11"',
        );
    });

    it('should just return task if StatusType has not changed', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const task = new TaskBuilder()
            .description('An already-DONE, non-recurring task 🏁 Delete')
            .dueDate(dueDate)
            .doneDate(dueDate)
            .status(Status.DONE)
            .build();
        expect(task.status.type).toEqual(StatusType.DONE);

        // Act
        const returnedTasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(returnedTasks.length).toEqual(1);
        expect(toLines(returnedTasks).join('\n')).toMatchInlineSnapshot(
            '"- [x] An already-DONE, non-recurring task 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10"',
        );
    });

    it('should just return trigger-less, non-recurring task', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const task = new TaskBuilder().description('A non-recurring task with no trigger').dueDate(dueDate).build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks.length).toEqual(1);
        expect(toLines(tasks).join('\n')).toMatchInlineSnapshot(
            '"- [x] A non-recurring task with no trigger 📅 2024-02-10 ✅ 2024-02-11"',
        );
    });

    it('should just return trigger-less recurring task', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const recurrence = new RecurrenceBuilder().rule('every day').dueDate(dueDate).build();
        const task = new TaskBuilder()
            .description('A recurring task with no trigger')
            .recurrence(recurrence)
            .dueDate(dueDate)
            .build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks.length).toEqual(2);
        expect(toLines(tasks).join('\n')).toMatchInlineSnapshot(`
            "- [ ] A recurring task with no trigger 🔁 every day 📅 2024-02-11
            - [x] A recurring task with no trigger 🔁 every day 📅 2024-02-10 ✅ 2024-02-11"
        `);
    });

    it('should return the task when going from TODO to IN_PROGRESS', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const recurrence = new RecurrenceBuilder().rule('every day').dueDate(dueDate).build();
        const task = new TaskBuilder()
            .description('A recurring task with 🏁 Delete')
            .recurrence(recurrence)
            .dueDate(dueDate)
            .build();

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeInProgress());

        // Assert
        expect(tasks.length).toEqual(1);
        expect(tasks[0].status.type).toEqual(StatusType.IN_PROGRESS);
    });

    it('should return the task when going from one DONE status to another DONE status', () => {
        // Arrange
        const done1 = new Status(new StatusConfiguration('x', 'done', ' ', true, StatusType.DONE));
        const done2 = new Status(new StatusConfiguration('X', 'DONE', ' ', true, StatusType.DONE));
        const task = new TaskBuilder().description('A simple done task with 🏁 Delete').status(done1).build();

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, done2);

        // Assert
        expect(tasks.length).toEqual(1);
        expect(tasks[0].status.symbol).toEqual('X');
        expect(tasks[0].status.type).toEqual(StatusType.DONE);
    });

    it('should return a task featuring the On Completion flag trigger but an empty string Action', () => {
        // Arrange
        const task = new TaskBuilder().description('A non-recurring task with 🏁').build();

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks.length).toEqual(1);
    });
});

type ToggleCase = {
    // inputs:
    nextStatus: Status;
    line: string;
};

function getCases(): ToggleCase[] {
    return [
        // Non-recurring
        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A non-recurring task with no trigger 📅 2024-02-10',
        },

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A non-recurring task with 🏁 Delete',
        },

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A non-recurring task with 🏁 Delete 📅 2024-02-10',
        },

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION 📅 2024-02-10',
        },

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A non-recurring task with 🏁',
        },

        // Recurring

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A recurring task with no trigger 🔁 every day 📅 2024-02-10',
        },

        {
            nextStatus: Status.makeDone(),
            line: '- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10',
        },

        {
            nextStatus: Status.makeInProgress(),
            line: '- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10',
        },

        // Other

        {
            nextStatus: Status.makeDone(),
            line: '- [x] An already-DONE task, changing to Same      DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10',
        },

        {
            nextStatus: new Status(new StatusConfiguration('X', 'new status', ' ', false, StatusType.DONE)),
            line: '- [x] An already-DONE task, changing to Different DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10',
        },
    ];
}

function action(toggleCase: ToggleCase): string {
    const newStatus = toggleCase.nextStatus;
    const task = fromLine({ line: toggleCase.line, path: 'anything.md', precedingHeader: 'heading' });
    const step1 = task.handleNewStatus(newStatus);
    const step2 = applyStatusAndOnCompletionAction(task, newStatus);
    return `
initial task:
${task.toFileLineString()}

=> advances to status [${newStatus.symbol}] and type ${newStatus.type}:
${toMarkdown(step1)}

=> which, after any on-completion action, results in:
${toMarkdown(step2)}
----------------------------------------------
`;
}

describe('visualise completion-behaviour', () => {
    it('visualise', () => {
        // List of status and task
        const cases = getCases();
        verifyAll('checking on completion', cases, (toggleCase) => action(toggleCase));
    });
});

describe('OnCompletion-Delete', () => {
    // TODO is there a better way to handle the following?  does an 'empty' Task exist?
    it('should return an empty Array for a non-recurring task with Delete Action', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const task = new TaskBuilder().description('A non-recurring task with 🏁 Delete').dueDate(dueDate).build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks).toEqual([]);
    });

    it('should return only the next instance of a recurring task with Delete action', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const recurrence = new RecurrenceBuilder().rule('every day').dueDate(dueDate).build();
        const task = new TaskBuilder()
            .description('A recurring task with 🏁 Delete')
            .recurrence(recurrence)
            .dueDate(dueDate)
            .build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks.length).toEqual(1);
        expect(toLines(tasks).join('\n')).toMatchInlineSnapshot(
            '"- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-11"',
        );
    });
    it('should delete a simple task with flag on completion', () => {
        // Arrange
        const task = new TaskBuilder().description('A non-recurring task with 🏁 Delete').build();

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks.length).toEqual(0);
    });
});

describe('OnCompletion-ToLogFile', () => {
    it('should write completed instance of non-recurring task to empty log file', () => {
        // Arrange
        const dueDate = '2024-02-10';
        const task = new TaskBuilder().description('A non-recurring task with 🏁 ToLogFile').dueDate(dueDate).build();
        expect(task.status.type).toEqual(StatusType.TODO);

        // Act
        const tasks = applyStatusAndOnCompletionAction(task, Status.makeDone());

        // Assert
        expect(tasks).toEqual([]);
    });
});

describe('OnCompletion-EndOfList', () => {
    it('should insert line at end of list when list is followed by a blank line', () => {
        const initialContentNoNewLine = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task

Final line in note file.`;
        const expectedContent = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST

Final line in note file.`;
        const targetListHeading = '## MY TASK LIST';
        const textToAppend = '- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST';
        const newFile = writeLineToListEnd(initialContentNoNewLine, targetListHeading, textToAppend);
        expect(newFile).toEqual(expectedContent);
    });

    it('should insert line at end of list when list followed by regular text', () => {
        const initialContentNoNewLine = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
A line of regular text.`;
        const expectedContent = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST
A line of regular text.`;
        const targetListHeading = '## MY TASK LIST';
        const textToAppend = '- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST';
        const newFile = writeLineToListEnd(initialContentNoNewLine, targetListHeading, textToAppend);
        expect(newFile).toEqual(expectedContent);
    });

    it('should insert line at end of list when list followed by a markdown heading', () => {
        const initialContentNoNewLine = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
## Another list heading`;
        const expectedContent = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST
## Another list heading`;
        const targetListHeading = '## MY TASK LIST';
        const textToAppend = '- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST';
        const newFile = writeLineToListEnd(initialContentNoNewLine, targetListHeading, textToAppend);
        expect(newFile).toEqual(expectedContent);
    });

    it('should insert line at end of list when end of list is last line in note file', () => {
        const initialContentNoNewLine = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task`;
        const expectedContent = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
- [ ] An incomplete task
- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST`;
        const targetListHeading = '## MY TASK LIST';
        const textToAppend = '- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST';
        const newFile = writeLineToListEnd(initialContentNoNewLine, targetListHeading, textToAppend);
        expect(newFile).toEqual(expectedContent);
    });

    it('should recognize indented tasks as part of list', () => {
        const initialContentNoNewLine = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
    - [ ] An INDENTED incomplete task`;
        const expectedContent = `Sed ipsam libero qui consequuntur quaerat non atque quia ab praesentium explicabo.
## MY TASK LIST
    - [ ] An INDENTED incomplete task
- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST`;
        const targetListHeading = '## MY TASK LIST';
        const textToAppend = '- [-] A COMPLETED TASK TO INSERT AT END OF NAMED LIST';
        const newFile = writeLineToListEnd(initialContentNoNewLine, targetListHeading, textToAppend);
        expect(newFile).toEqual(expectedContent);
    });
});
