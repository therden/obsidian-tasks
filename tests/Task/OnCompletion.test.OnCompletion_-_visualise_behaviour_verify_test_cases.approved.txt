checking on completion



initial task:
- [ ] A non-recurring task with no trigger 📅 2024-02-10

=> advances to status [x] and type DONE:
- [x] A non-recurring task with no trigger 📅 2024-02-10 ✅ 2024-02-11

=> which, after any on-completion action, results in:
- [x] A non-recurring task with no trigger 📅 2024-02-10 ✅ 2024-02-11
----------------------------------------------


initial task:
- [ ] A non-recurring task with 🏁 Delete

=> advances to status [x] and type DONE:
- [x] A non-recurring task with 🏁 Delete ✅ 2024-02-11

=> which, after any on-completion action, results in:

----------------------------------------------


initial task:
- [ ] A non-recurring task with 🏁 Delete 📅 2024-02-10

=> advances to status [x] and type DONE:
- [x] A non-recurring task with 🏁 Delete 📅 2024-02-10 ✅ 2024-02-11

=> which, after any on-completion action, results in:

----------------------------------------------


initial task:
- [ ] A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION 📅 2024-02-10

=> advances to status [x] and type DONE:
- [x] A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION 📅 2024-02-10 ✅ 2024-02-11

=> which, after any on-completion action, results in:
- [x] A non-recurring task with invalid OC trigger 🏁 INVALID_ACTION 📅 2024-02-10 ✅ 2024-02-11
----------------------------------------------


initial task:
- [ ] A non-recurring task with 🏁

=> advances to status [x] and type DONE:
- [x] A non-recurring task with 🏁 ✅ 2024-02-11

=> which, after any on-completion action, results in:
- [x] A non-recurring task with 🏁 ✅ 2024-02-11
----------------------------------------------


initial task:
- [ ] A recurring task with no trigger 🔁 every day 📅 2024-02-10

=> advances to status [x] and type DONE:
- [ ] A recurring task with no trigger 🔁 every day 📅 2024-02-11
- [x] A recurring task with no trigger 🔁 every day 📅 2024-02-10 ✅ 2024-02-11

=> which, after any on-completion action, results in:
- [ ] A recurring task with no trigger 🔁 every day 📅 2024-02-11
- [x] A recurring task with no trigger 🔁 every day 📅 2024-02-10 ✅ 2024-02-11
----------------------------------------------


initial task:
- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10

=> advances to status [x] and type DONE:
- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-11
- [x] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10 ✅ 2024-02-11

=> which, after any on-completion action, results in:
- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-11
----------------------------------------------


initial task:
- [ ] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10

=> advances to status [/] and type IN_PROGRESS:
- [/] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10

=> which, after any on-completion action, results in:
- [/] A recurring task with 🏁 Delete 🔁 every day 📅 2024-02-10
----------------------------------------------


initial task:
- [x] An already-DONE task, changing to Same      DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10

=> advances to status [x] and type DONE:
- [x] An already-DONE task, changing to Same      DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10

=> which, after any on-completion action, results in:
- [x] An already-DONE task, changing to Same      DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10
----------------------------------------------


initial task:
- [x] An already-DONE task, changing to Different DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10

=> advances to status [X] and type DONE:
- [X] An already-DONE task, changing to Different DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10

=> which, after any on-completion action, results in:
- [X] An already-DONE task, changing to Different DONE status 🏁 Delete 📅 2024-02-10 ✅ 2024-02-10
----------------------------------------------


initial task:
* [ ] Asterisk recurring task with 🏁 Delete 🔁 every day 📅 2024-04-29

=> advances to status [x] and type DONE:
* [ ] Asterisk recurring task with 🏁 Delete 🔁 every day 📅 2024-04-30
* [x] Asterisk recurring task with 🏁 Delete 🔁 every day 📅 2024-04-29 ✅ 2024-02-11

=> which, after any on-completion action, results in:
* [ ] Asterisk recurring task with 🏁 Delete 🔁 every day 📅 2024-04-30
----------------------------------------------


initial task:
+ [ ] Plus sign recurring task with 🏁 Delete 🔁 every day 📅 2024-04-30

=> advances to status [x] and type DONE:
+ [ ] Plus sign recurring task with 🏁 Delete 🔁 every day 📅 2024-05-01
+ [x] Plus sign recurring task with 🏁 Delete 🔁 every day 📅 2024-04-30 ✅ 2024-02-11

=> which, after any on-completion action, results in:
+ [ ] Plus sign recurring task with 🏁 Delete 🔁 every day 📅 2024-05-01
----------------------------------------------

